services/src/main/java/org/keycloak/authentication/forms/RegistrationProfile.java

```diff
@@ -67,33 +64,37 @@ public void validate(org.keycloak.authentication.ValidationContext context) {
 
         context.getEvent().detail(Details.REGISTER_METHOD, "form");
 
-        UserProfileValidationResult result = forRegistrationProfile(context.getSession(), formData).validate();
-        List<FormMessage> errors = Validation.getFormErrorsFromValidation(result);
+        UserProfileProvider profileProvider = context.getSession().getProvider(UserProfileProvider.class);
+        UserProfile profile = profileProvider.forValidation(UserProfileContext.REGISTRATION_PROFILE, formData);
+
+        try {
+            profile.validate();
```

Comment:
In default flow this runs just after `RegistrationUserCreation` action is executed, so I believe this validation is duplicate and only loads server and complicates stuff without real reason. But I can investigate this and patch it once this PR is accepted.

Reply:
It does not really complicates, right? Code is so simple :)

But I agree that it runs validation twice when actually creating/updating users. We can change that by setting the profile instance as an attribute in the session.